### PR TITLE
Fix #1708: duplicate symbols in package

### LINKED ISSUE #1708: crash on compilable code:at dotty.tools.dotc.core.Contexts$FreshContext.setOwner
scalac can compile it but dotc crashes.

$ cat abc.scala
```
package foo { trait identifier }
package foo { package identifier {} }
```

$ dotc abc.scala
```
exception occurred while typechecking abc.scala

exception occurred while compiling abc.scala
Exception in thread "main" java.lang.AssertionError: assertion failed
	at scala.Predef$.assert(Predef.scala:151)
	at dotty.tools.dotc.core.Contexts$FreshContext.setOwner(Contexts.scala:451)
	at dotty.tools.dotc.core.Contexts$Context.superOrThisCallContext(Contexts.scala:359)
	at dotty.tools.dotc.core.Contexts$Context.superCallContext(Contexts.scala:337)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1200)
	at dotty.tools.dotc.typer.Typer$$anonfun$typedClassDef$1.apply(Typer.scala:1198)
	at dotty.tools.dotc.util.Stats$.track(Stats.scala:35)
	at dotty.tools.dotc.typer.Typer.typedClassDef(Typer.scala:1198)
	at dotty.tools.dotc.typer.Typer.typedNamed$1(Typer.scala:1434)
	at dotty.tools.dotc.typer.Typer.typedUnadapted(Typer.scala:1482)
<snipped>
```

### PR DESCRIPTION
**Pull request by felixmulder**
review: @odersky

### CODE REVIEW

**Comment by felixmulder** [comment]
Fixed failing test, this error was not caught previously:

```
-- Error: tests/neg/i1643.scala --------------------------------------
1 |trait T extends Array {
  |      ^
  |      cannot extend final class Array

```

File: src/dotty/tools/dotc/typer/Namer.scala [hunk 1]
```diff
@@ -343,8 +344,17 @@ class Namer { typer: Typer =>
       case Select(qual: RefTree, _) => createPackageSymbol(qual).moduleClass
     }
     val existing = pkgOwner.info.decls.lookup(pid.name)
+
     if ((existing is Package) && (pkgOwner eq existing.owner)) existing
-    else ctx.newCompletePackageSymbol(pkgOwner, pid.name.asTermName).entered
+    else {
+      /** If there's already an existing type, then the package is a dup of this type */
+      val existingTpe = pkgOwner.info.decls.lookup(pid.name.toTypeName)
+      if (existingTpe != NoSymbol) {
```
Comment (odersky) [inline src/dotty/tools/dotc/typer/Namer.scala]: More idiomatic: `if (existingType.exists)`

File: src/dotty/tools/dotc/typer/Namer.scala [hunk 2]
```diff
@@ -343,8 +344,17 @@ class Namer { typer: Typer =>
...
+        ctx.error(PkgDuplicateSymbol(existingTpe), pid.pos)
+        ctx.newCompletePackageSymbol(pkgOwner, (pid.name ++ "$termDup").toTermName).entered
```
Comment (odersky) [inline src/dotty/tools/dotc/typer/Namer.scala]: I propose we use the naming scheme `pid.name ++ "$_error_` here.

File: src/dotty/tools/dotc/typer/Namer.scala [hunk 3]
```diff
@@ -261,10 +262,10 @@ class Namer { typer: Typer =>
         ctx.error(msg, tree.pos)
         name.freshened
       }
-      def preExisting = ctx.effectiveScope.lookup(name)
+      lazy val preExisting = ctx.effectiveScope.lookup(name)
```
Comment (odersky) [inline src/dotty/tools/dotc/typer/Namer.scala]: I don't think we need a lazy val here. preExisting is only used once on each path. Note that defs are more compact (in terms of byte code length) and efficient than lazy vals.

**Comment by felixmulder** [comment]
Done with suggested changes :+1:

**Comment by odersky** [comment]
LGTM, thanks!

**Pull request merged**